Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -909,28 +920,49 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
             notifyCursorChanged(toWebCoreCursorType(pdfElementTypes, altKeyIsActive));
             return true;
         }
+        case WebMouseEventButton::Left: {
+            if (m_trackedAnnotation && m_trackedAnnotation != annotationForRootViewPoint(event.position())) {
+                handleMouseDraggedOffTrackedAnnotation();
+                return true;
+            }
+            return false;
+        }
         default:
             return false;
         }
     case WebEventType::MouseDown:
         switch (event.button()) {
         case WebMouseEventButton::Left: {
-            auto pointInDocumentSpace = convertFromPluginToDocument(convertFromRootViewToPlugin(event.position()));
-            auto nearestPageIndex = nearestPageIndexForDocumentPoint(pointInDocumentSpace);
-            if (!nearestPageIndex)
+            if (RetainPtr<PDFAnnotation> annotation = annotationForRootViewPoint(event.position())) {
+                if (([annotation isKindOfClass:getPDFAnnotationButtonWidgetClass()] || [annotation isKindOfClass:getPDFAnnotationTextWidgetClass()] || [annotation isKindOfClass:getPDFAnnotationChoiceWidgetClass()]) && [annotation isReadOnly])
+                    return true;
+
+                if ([annotation isKindOfClass:getPDFAnnotationTextWidgetClass()] || [annotation isKindOfClass:getPDFAnnotationChoiceWidgetClass()]) {
+                    setActivePDFPluginAnnotation(annotation.get());
+                    return true;
+                }
+                if ([annotation isKindOfClass:getPDFAnnotationButtonWidgetClass()]) {
+                    startAnnotationTracking(annotation);
+                    return true;
+                }
                 return false;
-
-            auto page = m_documentLayout.pageAtIndex(nearestPageIndex.value());
-            auto pointInPDFPageSpace = convertFromDocumentToPage(pointInDocumentSpace, nearestPageIndex.value());
-            if (auto annotation = [page annotationAtPoint:pointInPDFPageSpace]; annotation && [annotation isKindOfClass:getPDFAnnotationTextWidgetClass()]) {
-                setActiveAnnotation(annotation);
-                return true;
             }
             return false;
         }
         default:
             return false;
         }
+    case WebEventType::MouseUp:
```

Comment:
What if you click in an iframe PDF and then drag out. Do we miss the MouseUp? If we do, does it leave us in a bad state?

Reply:
It does seem to leave us in a bad state (even happens with main frame PDFs on shipping)! To help get us out of this bad state, I added some logic that will check if there is a tracked annotation on a mouse move and treat it as if the cursor was dragged off the annotation